File: crawler.py
#!/usr/bin/env python3
import os
import sys
import zipfile

import requests
import structlog
from bs4 import BeautifulSoup

DRIVER_DOWNLOAD_URL = "https://cloud.google.com/bigquery/docs/reference/odbc-jdbc-drivers"
DRIVER_FILENAME = "GoogleBigQueryJDBC42.jar"

EXCLUDE_DRIVER_LIST = [
    "SimbaJDBCDriverforGoogleBigQuery42_1.5.4.1008.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.5.0.1001.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.3.3.1004.zip",
    "SimbaBigQueryJDBC42-1.3.2.1003.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.3.0.1001.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.25.1029.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.23.1027.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.22.1026.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.21.1025.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.19.1023.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.18.1022.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.16.1020.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.14.1017.zip",
    "SimbaJDBCDriverforGoogleBigQuery42_1.2.1.1001.zip",
    "SimbaJDBCDriverforGoogleBigQuery41_1.2.1.1001.zip",
]

# Get logger from structlog
logger = structlog.get_logger()


def fetch_page_content(url: str) -> str | None:
    response = requests.get(url)
    if response.status_code == 200:
        return response.text
    return None


def get_driver_download_links(page_content: str) -> list[str]:
    soup = BeautifulSoup(page_content, "html.parser")
    links = []
    for download_link in soup.find_all("a", href=True):
        if "jdbc" in download_link["href"]:
            links.append(download_link["href"])
    return links


def exclude_old_drivers(driver_links: list[str]) -> list[str]:
    # Extract the filename from the URL
    def extract_filename(l):
        return l.split("/")[-1]

    return [l for l in driver_links if extract_filename(l) not in EXCLUDE_DRIVER_LIST]


def download_jdbc_driver(driver_link: str, dest_dir: str = "downloads") -> bool:
    response = requests.get(driver_link)
    if response.status_code == 200:
        f = f"{dest_dir}/{driver_link.split('/')[-1]}"
        with open(f, "wb") as file:
            file.write(response.content)
        if zipfile.is_zipfile(f):
            return True
        else:
            logger.error(f"{f} is not a zip file.")
            return False
    return False


def extract_specific_jar(zip_path: str, extract_to: str = "downloads") -> bool:
    jar_name = DRIVER_FILENAME
    zip_file_name = ".".join(zip_path.split("/")[-1].split(".")[0:-1])

    with zipfile.ZipFile(zip_path, "r") as zip_ref:
        all_files = zip_ref.namelist()
        if jar_name not in all_files:
            logger.warn(f"{jar_name} not found in the zip file.")
            return False

        # Extract the specific jar file
        zip_ref.extract(jar_name, extract_to)
        os.rename(f"{extract_to}/{jar_name}", f"{extract_to}/{zip_file_name}-{jar_name}")
        logger.info(f"Extracted {jar_name} to {extract_to}")
        return True


def check_downloaded_file(driver_link: str) -> bool:
    with open("download_history.txt") as file:
        lines = [l.strip() for l in file.readlines()]
        if driver_link in lines:
            return True
    return False


def append_to_download_history(driver_link: str) -> None:
    if check_downloaded_file(driver_link):
        return

    with open("download_history.txt", "a") as file:
        file.write(driver_link + "\n")


if __name__ == "__main__":
    page_content = fetch_page_content(DRIVER_DOWNLOAD_URL)
    if page_content is None:
        logger.error("Failed to fetch the page content")
        sys.exit(1)

    jdbc_links = get_driver_download_links(page_content)
    for link in exclude_old_drivers(jdbc_links):
        if check_downloaded_file(link):
            logger.info(f"{link} has already been downloaded.")
            continue

        if link.endswith(".zip"):
            logger.debug(link)
            result = download_jdbc_driver(link)
            if not result:
                logger.error(f"Failed to download {link}")
                sys.exit(1)
            logger.info(f"Downloaded {link}")

            # Extract the specific jar file
            extract_result = extract_specific_jar(f"downloads/{link.split('/')[-1]}")
            if not extract_result:
                logger.error(f"Failed to extract the jar file from {link}")
                sys.exit(1)
            append_to_download_history(link)
            logger.info(f"Extracted {DRIVER_FILENAME} from {link}")

File: crawler.go
package main

import (
	"archive/zip"
	"bufio"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

const (
	driverDownloadURL = "https://cloud.google.com/bigquery/docs/reference/odbc-jdbc-drivers"
	driverFilename    = "GoogleBigQueryJDBC42.jar"
	downloadsDir      = "downloads"
	historyFile       = "download_history.txt"
	httpTimeout       = 180 * time.Second
)

var excludeDriverList = map[string]struct{}{
	"SimbaJDBCDriverforGoogleBigQuery42_1.5.4.1008.zip":  {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.5.0.1001.zip":  {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.3.3.1004.zip":  {},
	"SimbaBigQueryJDBC42-1.3.2.1003.zip":                 {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.3.0.1001.zip":  {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.25.1029.zip": {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.23.1027.zip": {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.22.1026.zip": {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.21.1025.zip": {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.19.1023.zip": {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.18.1022.zip": {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.16.1020.zip": {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.14.1017.zip": {},
	"SimbaJDBCDriverforGoogleBigQuery42_1.2.1.1001.zip":  {},
	"SimbaJDBCDriverforGoogleBigQuery41_1.2.1.1001.zip":  {},
}

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	fmt.Println("Fetching driver list page...")
	page, err := fetchPageContent(driverDownloadURL)
	if err != nil {
		return fmt.Errorf("ページ取得に失敗しました: %w", err)
	}

	links := getDriverDownloadLinks(page)
	for i := range links {
		links[i] = normalizeURL(links[i])
	}
	links = excludeOldDrivers(links)

	// Ensure downloads directory exists
	if err := os.MkdirAll(downloadsDir, 0o755); err != nil {
		return fmt.Errorf("出力ディレクトリ作成に失敗: %w", err)
	}

	for _, link := range links {
		if !strings.HasSuffix(strings.ToLower(link), ".zip") {
			continue
		}
		already, err := isDownloaded(link)
		if err != nil {
			return err
		}
		if already {
			fmt.Printf("already downloaded: %s\n", link)
			continue
		}

		fmt.Printf("downloading: %s\n", link)
		zipPath, err := downloadJDBCDriver(link, downloadsDir)
		if err != nil {
			return fmt.Errorf("ダウンロード失敗: %s: %w", link, err)
		}
		fmt.Printf("downloaded: %s\n", zipPath)

		if err := extractSpecificJar(zipPath, downloadsDir); err != nil {
			return fmt.Errorf("JAR 抽出失敗 (%s): %w", link, err)
		}
		if err := appendToHistory(link); err != nil {
			return err
		}
		fmt.Printf("extracted: %s\n", driverFilename)
	}

	return nil
}

func httpClient() *http.Client {
	return &http.Client{Timeout: httpTimeout}
}

func fetchPageContent(url string) (string, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return "", err
	}
	resp, err := httpClient().Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("HTTP status %d", resp.StatusCode)
	}
	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// getDriverDownloadLinks extracts href values that contain "jdbc".
// For simplicity and zero external deps, use a regex on anchor tags.
func getDriverDownloadLinks(html string) []string {
	// Matches href="..." and captures the value; case-insensitive search for jdbc in the value.
	re := regexp.MustCompile(`(?i)<a[^>]+href\s*=\s*"([^"]+)"[^>]*>`) // "
	matches := re.FindAllStringSubmatch(html, -1)
	var out []string
	for _, m := range matches {
		if len(m) >= 2 {
			href := m[1]
			if strings.Contains(strings.ToLower(href), "jdbc") {
				out = append(out, href)
			}
		}
	}
	return out
}

func normalizeURL(href string) string {
	// 既に絶対 URL ならそのまま
	low := strings.ToLower(href)
	if strings.HasPrefix(low, "http://") || strings.HasPrefix(low, "https://") {
		return href
	}
	// 先頭がスラッシュの相対 URL のみベースドメインを補完
	if strings.HasPrefix(href, "/") {
		return "https://cloud.google.com" + href
	}
	return href
}

func excludeOldDrivers(links []string) []string {
	var out []string
	for _, l := range links {
		base := filepath.Base(l)
		if _, found := excludeDriverList[base]; found {
			continue
		}
		out = append(out, l)
	}
	return out
}

func downloadJDBCDriver(url, destDir string) (string, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return "", err
	}
	resp, err := httpClient().Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("HTTP status %d", resp.StatusCode)
	}

	filename := filepath.Base(url)
	path := filepath.Join(destDir, filename)
	f, err := os.Create(path)
	if err != nil {
		return "", err
	}
	defer f.Close()
	if _, err := io.Copy(f, resp.Body); err != nil {
		return "", err
	}

	// Validate zip by trying to open.
	if err := validateZip(path); err != nil {
		return "", err
	}
	return path, nil
}

func validateZip(path string) error {
	fi, err := os.Stat(path)
	if err != nil {
		return err
	}
	r, err := zip.OpenReader(path)
	if err != nil {
		return err
	}
	// Ensure close
	_ = r.Close()
	if fi.Size() == 0 {
		return errors.New("downloaded file is empty")
	}
	return nil
}

func extractSpecificJar(zipPath, extractTo string) error {
	zr, err := zip.OpenReader(zipPath)
	if err != nil {
		return err
	}
	defer zr.Close()

	var jarFile *zip.File
	for _, f := range zr.File {
		if f.Name == driverFilename {
			jarFile = f
			break
		}
	}
	if jarFile == nil {
		return fmt.Errorf("%s が ZIP 内に見つかりません", driverFilename)
	}

	rc, err := jarFile.Open()
	if err != nil {
		return err
	}
	defer rc.Close()

	base := strings.TrimSuffix(filepath.Base(zipPath), filepath.Ext(zipPath))
	outName := fmt.Sprintf("%s-%s", base, driverFilename)
	outPath := filepath.Join(extractTo, outName)

	outFile, err := os.Create(outPath)
	if err != nil {
		return err
	}
	defer outFile.Close()

	if _, err := io.Copy(outFile, rc); err != nil {
		return err
	}
	return nil
}

func isDownloaded(link string) (bool, error) {
	f, err := os.Open(historyFile)
	if err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}
		return false, err
	}
	defer f.Close()

	s := bufio.NewScanner(f)
	for s.Scan() {
		if strings.TrimSpace(s.Text()) == link {
			return true, nil
		}
	}
	if err := s.Err(); err != nil {
		return false, err
	}
	return false, nil
}

func appendToHistory(link string) error {
	already, err := isDownloaded(link)
	if err != nil {
		return err
	}
	if already {
		return nil
	}
	f, err := os.OpenFile(historyFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return err
	}
	defer f.Close()
	if _, err := f.WriteString(link + "\n"); err != nil {
		return err
	}
	return nil
}
